File: java/formula-java/src/main/java/org/formula/FormulaParser.java
package org.formula;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import java.util.function.Consumer;
import org.formula.context.DataContext;
import org.formula.parse.shuntingyard.Associativity;
import org.formula.parse.shuntingyard.Operator1;
import org.formula.parse.shuntingyard.Operator2;
import org.formula.parse.shuntingyard.ShuntingYardParser;
import org.formula.util.Ordinal;

public class FormulaParser {
    private final ShuntingYardParser parser = ShuntingYardParser.create();

    public static FormulaParser createDefault() {
        return new FormulaParser()
                .addNulls()
                .addLists()
                .addDefaultVariables()
                .addDefaultComments()
                .addBasicMath()
                .addBasicLogic()
                .addStringFunctions();
    }

    public FormulaParser addCustom(Consumer<ShuntingYardParser> modifyParserFn) {
        modifyParserFn.accept(parser);
        return this;
    }

    public Resolvable parse(String formulaText) {
        if (formulaText.isBlank()) {
            return Resolvable.empty();
        }
        return parser.parse(formulaText);
    }

    public FormulaParser addBasicMath() {
        parser
                .operator("^", 4, Associativity.RIGHT, (a, b) -> ResolvedValue.of(Math.pow(a.asDecimal(), b.asDecimal())))
                .operator("*", 3, Associativity.LEFT, (a, b) -> ResolvedValue.of(a.asDecimal() * b.asDecimal()))
                .operator("/", 3, Associativity.LEFT, (a, b) -> ResolvedValue.of(a.asDecimal() / b.asDecimal()))
                .operator("+", 2, Associativity.LEFT, FormulaParser::addReduceFn)
                .biOperator("-",
                        new Operator1("-", 4, Associativity.LEFT, a -> ResolvedValue.of(-a.asDecimal())),
                        new Operator2("-", 2, Associativity.LEFT, (a, b) -> ResolvedValue.of(a.asDecimal() - b.asDecimal())))
                .function("abs", (ResolvedValue a) -> ResolvedValue.of(Math.abs(a.asDecimal())))
                .function("min", FormulaParser::minFn)
                .function("max", FormulaParser::maxFn)
                .function("maxeach", FormulaParser::maxEachFn)
                .function("mineach", FormulaParser::minEachFn)
                .function("clamp", FormulaParser::clampFn)
                .function("floor", (ResolvedValue a) -> ResolvedValue.of(Math.floor(a.asDecimal())))
                .function("ceil", (ResolvedValue a) -> ResolvedValue.of(Math.ceil(a.asDecimal())))
                .function("sum", FormulaParser::sumFn);
        return this;
    }

    public FormulaParser addBasicLogic() {
        this.parser
                .operator("!", 2, Associativity.LEFT, (ResolvedValue a) -> ResolvedValue.of(!a.asBoolean()))
                .operator("<", 3, Associativity.LEFT, (ResolvedValue a, ResolvedValue b) -> ResolvedValue.of(a.asDecimal() < b.asDecimal()))
                .operator("<=", 3, Associativity.LEFT, (ResolvedValue a, ResolvedValue b) -> ResolvedValue.of(a.asDecimal() <= b.asDecimal()))
                .operator(">", 3, Associativity.LEFT, (ResolvedValue a, ResolvedValue b) -> ResolvedValue.of(a.asDecimal() > b.asDecimal()))
                .operator(">=", 3, Associativity.LEFT, (ResolvedValue a, ResolvedValue b) -> ResolvedValue.of(a.asDecimal() >= b.asDecimal()))
                .operator("==", 3, Associativity.LEFT, (ResolvedValue a, ResolvedValue b) -> ResolvedValue.of(a.equals(b)))
                .operator("!=", 3, Associativity.LEFT, (ResolvedValue a, ResolvedValue b) -> ResolvedValue.of(!a.equals(b)))
                .operator("AND", 1, Associativity.LEFT, (ResolvedValue a, ResolvedValue b) -> ResolvedValue.of(a.asBoolean() && b.asBoolean()))
                .operator("OR", 1, Associativity.LEFT, (ResolvedValue a, ResolvedValue b) -> ResolvedValue.of(a.asBoolean() || b.asBoolean()))
                .term("true", () -> ResolvedValue.TRUE)
                .term("false", () -> ResolvedValue.FALSE)
                .function("if", (ResolvedValue a, ResolvedValue b, ResolvedValue c) -> a.asBoolean() ? b : c)
                .function("any", FormulaParser::anyFn)
                .function("all", FormulaParser::allFn);
        return this;
    }

    public FormulaParser addDefaultVariables() {
        this.parser
                .variable("@", FormulaParser::variableFn)
                .variable("@{", "}", FormulaParser::variableFn);
        return this;
    }

    public FormulaParser addDefaultComments() {
        this.parser
                .comment("[", "]", (value, comment) -> NamedResolvedValue.of(value, comment.substring(1, comment.length() - 1), "[", "]"));
        return this;
    }

    public FormulaParser addNulls() {
        this.parser
                .term("null", ResolvedValue::none);
        return this;
    }

    public FormulaParser addLists() {
        this.parser
                .operator(",", 1, Associativity.LEFT, FormulaParser::mergeLists);
        return this;
    }

    public FormulaParser addStringFunctions() {
        this.parser
                .function("signed", (ResolvedValue a) -> ResolvedValue.of((a.asNumber() < 0 ? "" : "+") + a.asNumber()))

                .function("concat", FormulaParser::concatFn)
                .function("ordinal", (ResolvedValue a) -> ResolvedValue.of(Ordinal.toString(a.asNumber())));
        return this;
    }

    private static ResolvedValue allFn(List<ResolvedValue> values) {
        for (ResolvedValue value : values) {
            if (!value.hasValue()) return ResolvedValue.FALSE;
            var valueAsList = value.asList();
            for (ResolvedValue el : valueAsList) {
                if (!el.hasValue() || !el.asBoolean()) {
                    return ResolvedValue.FALSE;
                }
            }
        }
        return ResolvedValue.TRUE;
    }

    private static ResolvedValue anyFn(List<ResolvedValue> values) {
        for (var arg : values) {
            if (!arg.hasValue()) {
                continue;
            }
            for (var el : arg.asList()) {
                if (el.hasValue() && el.asBoolean()) return ResolvedValue.TRUE;
            }
        }
        return ResolvedValue.FALSE;
    }

    private static ResolvedValue sumFn(ResolvedValue a) {
        var list = a.asList();
        if (list.isEmpty()) {
            return ResolvedValue.ZERO;
        }
        if (list.size() < 2) {
            return a;
        }
        return ResolvedValue.of(a.asList()
                .stream()
                .mapToDouble(value -> sumFn(value).asDecimal())
                .sum());
    }

    private static ResolvedValue maxEachFn(ResolvedValue a) {
        List<ResolvedValue> values = new ArrayList<>();
        a.asList().forEach(value -> values.add(maxFn(value)));
        return ResolvedValue.of(values);
    }

    private static ResolvedValue minEachFn(ResolvedValue a) {
        return ResolvedValue.of(a.asList().stream()
                .map(FormulaParser::minFn)
                .toList());
    }

    private static ResolvedValue clampFn(ResolvedValue a, ResolvedValue min, ResolvedValue max) {
        if (a.asDecimal() < min.asDecimal()) {
            return min;
        }
        if (a.asDecimal() > max.asDecimal()) {
            return max;
        }
        return a;
    }

    private static ResolvedValue maxFn(ResolvedValue a) {
        return a.asList().stream()
                .max(Comparator.comparing(ResolvedValue::asDecimal))
                .orElse(ResolvedValue.none());
    }

    private static ResolvedValue minFn(ResolvedValue a) {
        return a.asList().stream()
                .min(Comparator.comparing(ResolvedValue::asDecimal))
                .orElse(ResolvedValue.none());
    }

    private static ResolvedValue mergeLists(ResolvedValue a, ResolvedValue b) {
        List<ResolvedValue> merged = new ArrayList<>();
        merged.addAll(a.hasValue() ? a.asList() : List.of(ResolvedValue.none()));
        merged.addAll(b.hasValue() ? b.asList() : List.of(ResolvedValue.none()));
        return ResolvedValue.of(merged);
    }

    private static ResolvedValue variableFn(DataContext context, String key) {
        if (key.contains("*")) {
            return ResolvedValue.of(context.search(key).toList());
        }
        return context.get(key);
    }

    private static ResolvedValue concatFn(List<ResolvedValue> values) {
        return ResolvedValue.concat(values);
    }

    private static ResolvedValue addReduceFn(ResolvedValue a, ResolvedValue b) {
        if (!a.hasValue() && !b.hasValue()) {
            return ResolvedValue.ZERO;
        }
        return ResolvedValue.of(a.asDecimal() + b.asDecimal());
    }

    private static ResolvedValue maxReduceFn(ResolvedValue a, ResolvedValue b) {
        return checkForNone(a,b).orElseGet(() -> a.asDecimal() > b.asDecimal() ? a : b);
    }

    private static ResolvedValue minReduceFn(ResolvedValue a, ResolvedValue b) {
        return checkForNone(a,b).orElseGet(() -> a.asDecimal() < b.asDecimal() ? a : b);
    }

    private static Optional<ResolvedValue> checkForNone(ResolvedValue a, ResolvedValue b) {
        if (!a.hasValue()) return Optional.of(b);
        if (!b.hasValue()) return Optional.of(a);
        return Optional.empty();
    }

    private FormulaParser() {}
}

File: js/formula-js/src/FormulaParser.ts
import {Associativity, ShuntingYard, ShuntingYardParser} from "./ShuntingYard";
import {ResolvedValue} from "./ResolvedValue";
import {DataContext} from "./DataContext";
import {NamedResolvedValue} from "./NamedResolvedValue";
import {Resolvable} from "./Resolvable";

export class FormulaParser {
  private readonly parser = ShuntingYard.parser();

  static createDefault(): FormulaParser {
    return new FormulaParser()
    .addNulls()
    .addMathOperators()
    .addMathFunctions()
    .addLogicOperators()
    .addLogicFunctions()
    .addLists()
    .addVariables()
    .addOrdinalFunctions()
    .addStringFunctions()
    .addComments();
  }

  constructor() {
  }

  public addCustom(modifyParserFn: (parser: ShuntingYardParser) => void): FormulaParser {
    modifyParserFn(this.parser);
    return this;
  }

  public parse(formula: string): Resolvable {
    return this.parser.parse(formula);
  }

  public addNulls(): FormulaParser {
    this.parser
    .term('null', () => ResolvedValue.None);
    return this;
  }

  public addMathOperators(): FormulaParser {
    this.parser
    .biOperator('-',
        4, Associativity.Left, (a: ResolvedValue) => ResolvedValue.of(-a.asNumber()),
        2, Associativity.Left, (a: ResolvedValue, b: ResolvedValue) => ResolvedValue.of(a.asNumber() - b.asNumber()))
    .operator('^', 4, Associativity.Right, 2, (a: ResolvedValue, b: ResolvedValue) => ResolvedValue.of(Math.pow(a.asNumber(), b.asNumber())))
    .operator('*', 3, Associativity.Left, 2, (a: ResolvedValue, b: ResolvedValue) => ResolvedValue.of(a.asNumber() * b.asNumber()))
    .operator('/', 3, Associativity.Left, 2, (a: ResolvedValue, b: ResolvedValue) => ResolvedValue.of(a.asNumber() / b.asNumber()))
    .operator('+', 2, Associativity.Left, 2, FormulaParser.addReduceFn)
    .operator('!', 2, Associativity.Left, 1, (a: ResolvedValue) => ResolvedValue.of(!a.asBoolean()))
    .operator('<', 3, Associativity.Left, 2, (a: ResolvedValue, b: ResolvedValue) => ResolvedValue.of(a.asNumber() < b.asNumber()))
    .operator('<=', 3, Associativity.Left, 2, (a: ResolvedValue, b: ResolvedValue) => ResolvedValue.of(a.asNumber() <= b.asNumber()))
    .operator('>', 3, Associativity.Left, 2, (a: ResolvedValue, b: ResolvedValue) => ResolvedValue.of(a.asNumber() > b.asNumber()))
    .operator('>=', 3, Associativity.Left, 2, (a: ResolvedValue, b: ResolvedValue) => ResolvedValue.of(a.asNumber() >= b.asNumber()))
    .operator('==', 3, Associativity.Left, 2, (a: ResolvedValue, b: ResolvedValue) => ResolvedValue.of(a.equals(b)))
    .operator('!=', 3, Associativity.Left, 2, (a: ResolvedValue, b: ResolvedValue) => ResolvedValue.of(!a.equals(b)))
    return this;
  }

  public addMathFunctions(): FormulaParser {
    this.parser
    .function('abs', 1, (a: ResolvedValue) => ResolvedValue.of(Math.abs(a.asNumber())))
    .function('floor', 1, (a: ResolvedValue) => ResolvedValue.of(Math.floor(a.asNumber())))
    .function('ceil', 1, (a: ResolvedValue) => ResolvedValue.of(Math.ceil(a.asNumber())))
    .function('signed', 1, (a: ResolvedValue) => ResolvedValue.of((a.asNumber() < 0 ? '' : '+') + a.asNumber()))
    .function('sum', 1, FormulaParser.sumFn)
    .function('clamp', 3, FormulaParser.clampFn)
    .function('max', 1, FormulaParser.maxFn)
    .function('maxeach', 1, FormulaParser.maxEachFn)
    .function('mineach', 1, FormulaParser.minEachFn)
    .function('min', 1, FormulaParser.minFn)
    return this;
  }

  public addLogicOperators(): FormulaParser {
    this.parser
    .operator('AND', 1, Associativity.Left, 2, (a: ResolvedValue, b: ResolvedValue) => ResolvedValue.of(a.asBoolean() && b.asBoolean()))
    .operator('OR', 1, Associativity.Left, 2, (a: ResolvedValue, b: ResolvedValue) => ResolvedValue.of(a.asBoolean() || b.asBoolean()))
    .term('true', () => ResolvedValue.of(true))
    .term('false', () => ResolvedValue.of(false))
    return this;
  }

  public addLogicFunctions(): FormulaParser {
    this.parser
    .function('if', 3, (a: ResolvedValue, b: ResolvedValue, c: ResolvedValue) => a.asBoolean() ? b : c)
    .varargsFunction('any', FormulaParser.anyFn)
    .varargsFunction('all', FormulaParser.allFn);
    return this;
  }

  public addOrdinalFunctions(): FormulaParser {
    this.parser
    .function('ordinal', 1, (a: ResolvedValue) => ResolvedValue.of(ordinal(a.asNumber())));
    return this;
  }

  public addStringFunctions(): FormulaParser {
    this.parser
    .varargsFunction('concat', FormulaParser.concatFn);
    return this;
  }

  public addVariables(): FormulaParser {
    this.parser
    .variable('@', '', FormulaParser.variableFn)
    .variable('@{', '}', FormulaParser.variableFn);
    return this;
  }

  public addComments(): FormulaParser {
    this.parser
    .comment('[', ']', (text, value) => new NamedResolvedValue(value, text));
    return this;
  }

  public addLists(): FormulaParser {
    this.parser
    .operator(',', 1, Associativity.Left, 2, FormulaParser.mergeLists);
    return this;
  }

  private static variableFn(state: DataContext, key: string) {
    if (key.includes('*')) {
      return ResolvedValue.of(state.search(key));
    }
    return state.get(key);
  }

  private static minFn(a: ResolvedValue): ResolvedValue {
    const list = a.asList();
    if (list.length < 2) return a;
    return list
    .reduce((a, b) => a.asNumber() < b.asNumber() ? a : b);
  }

  private static maxFn(a: ResolvedValue): ResolvedValue {
    const list = a.asList();
    if (list.length < 2) return a;
    return list
    .reduce((a, b) => a.asNumber() > b.asNumber() ? a : b);
  }

  private static maxEachFn(a: ResolvedValue): ResolvedValue {
    return ResolvedValue.of(a.asList()
    .map(FormulaParser.maxFn));
  }

  private static minEachFn(a: ResolvedValue): ResolvedValue {
    return ResolvedValue.of(a.asList()
    .map(FormulaParser.minFn));
  }

  private static sumFn(a: ResolvedValue) {
    const list = a.asList();
    if (list.length === 0) return ResolvedValue.of(0);
    if (list.length < 2) return a;
    return ResolvedValue.of(list
    .reduce((a, b) => a + FormulaParser.sumFn(b).asNumber(), 0));
  }

  private static concatFn(args: ResolvedValue[]) {
    return ResolvedValue.of(
        args.flatMap(a => a.asList()));
  }

  private static addReduceFn(a: ResolvedValue, b: ResolvedValue): ResolvedValue {
    if ((a ?? ResolvedValue.None).equals(ResolvedValue.None) && (b ?? ResolvedValue.None).equals(ResolvedValue.None)) return ResolvedValue.None;
    return ResolvedValue.of(a.asNumber() + b.asNumber());
  }

  private static clampFn(a: ResolvedValue, min: ResolvedValue, max: ResolvedValue) {
    return ResolvedValue.of(Math.min(max.asNumber(), Math.max(min.asNumber(), a.asNumber())));
  }

  private static allFn(args: ResolvedValue[]): ResolvedValue {
    for (const arg of args) {
      if (!arg.hasValue()) return ResolvedValue.False;
      const argList = arg.asList();
      for (const el of argList) {
        if (!el.hasValue() || !el.asBoolean()) return ResolvedValue.False;
      }
    }
    return ResolvedValue.True;
  }

  private static anyFn(args: ResolvedValue[]): ResolvedValue {
    for (const arg of args) {
      if (!arg.hasValue()) continue;
      for (const el of arg.asList()) {
        if (el.hasValue() && el.asBoolean()) return ResolvedValue.True;
      }
    }
    return ResolvedValue.False;
  }

  private static mergeLists(a: ResolvedValue, b: ResolvedValue) {
    return ResolvedValue.of([
      ...(a.hasValue() ? a.asList() : [ResolvedValue.None]),
      ...(b.hasValue() ? b.asList() : [ResolvedValue.None])
    ])
  }
}

function ordinal(n: number): string {
  let s = ["th", "st", "nd", "rd"];
  let v = n % 100;
  return n + (s[(v - 20) % 10] || s[v] || s[0]);
}